File: backend-node/src/services/emailService.ts
import jwt from 'jsonwebtoken';
import { config } from 'dotenv';

config();

/**
 * Email Service - NoopEmailSender (Placeholder)
 *
 * IMPORTANT: This is a placeholder implementation for development.
 * In production, replace this with a real email service provider like:
 * - SendGrid
 * - AWS SES
 * - Mailgun
 * - Postmark
 * - etc.
 *
 * For now, this service logs confirmation links to the console instead
 * of sending actual emails.
 */

const JWT_SECRET = process.env.JWT_SECRET || 'your-super-secret-jwt-key';
const FRONTEND_URL = process.env.FRONTEND_URL || 'http://localhost:3000';

/**
 * Generate email confirmation token
 * Creates a JWT token that can be used to confirm a user's email address
 */
export function generateConfirmationToken(userId: string): string {
  // Token expires in 24 hours
  return jwt.sign(
    { userId, type: 'email-confirmation' },
    JWT_SECRET,
    { expiresIn: '24h' }
  );
}

/**
 * Verify email confirmation token
 * Validates the token and returns the user ID if valid
 */
export function verifyConfirmationToken(token: string): { userId: string } {
  try {
    const decoded = jwt.verify(token, JWT_SECRET) as { userId: string; type: string };

    if (decoded.type !== 'email-confirmation') {
      throw new Error('Invalid token type');
    }

    return { userId: decoded.userId };
  } catch (error) {
    if (error instanceof jwt.TokenExpiredError) {
      throw new Error('Confirmation token has expired');
    }
    if (error instanceof jwt.JsonWebTokenError) {
      throw new Error('Invalid confirmation token');
    }
    throw error;
  }
}

/**
 * Send email confirmation email
 *
 * PLACEHOLDER: In production, this would send an actual email.
 * For now, it logs the confirmation link to the console.
 */
export async function sendConfirmationEmail(email: string, token: string): Promise<void> {
  const confirmationLink = `${FRONTEND_URL}/confirm-email?token=${token}`;

  console.log('\n========================================');
  console.log('EMAIL CONFIRMATION (Development Mode)');
  console.log('========================================');
  console.log(`To: ${email}`);
  console.log(`Subject: Confirm your email address`);
  console.log(`\nConfirmation Link: ${confirmationLink}`);
  console.log('========================================\n');

  // TODO: In production, replace with actual email sending logic:
  // await emailProvider.send({
  //   to: email,
  //   subject: 'Confirm your email address',
  //   html: `<p>Click this link to confirm your email: <a href="${confirmationLink}">${confirmationLink}</a></p>`,
  // });
}

/**
 * Send password reset email
 *
 * PLACEHOLDER: In production, this would send an actual email.
 * For now, it logs the reset link to the console.
 */
export async function sendPasswordResetEmail(email: string, token: string): Promise<void> {
  const resetLink = `${FRONTEND_URL}/reset-password?token=${token}`;

  console.log('\n========================================');
  console.log('PASSWORD RESET (Development Mode)');
  console.log('========================================');
  console.log(`To: ${email}`);
  console.log(`Subject: Reset your password`);
  console.log(`\nReset Link: ${resetLink}`);
  console.log('========================================\n');

  // TODO: In production, replace with actual email sending logic:
  // await emailProvider.send({
  //   to: email,
  //   subject: 'Reset your password',
  //   html: `<p>Click this link to reset your password: <a href="${resetLink}">${resetLink}</a></p>`,
  // });
}

/**
 * Send welcome email
 *
 * PLACEHOLDER: In production, this would send an actual email.
 * For now, it logs the welcome message to the console.
 */
export async function sendWelcomeEmail(email: string, username: string): Promise<void> {
  console.log('\n========================================');
  console.log('WELCOME EMAIL (Development Mode)');
  console.log('========================================');
  console.log(`To: ${email}`);
  console.log(`Subject: Welcome to DuoStackDemo!`);
  console.log(`\nHello ${username},\nWelcome to DuoStackDemo! We're excited to have you on board.`);
  console.log('========================================\n');

  // TODO: In production, replace with actual email sending logic:
  // await emailProvider.send({
  //   to: email,
  //   subject: 'Welcome to DuoStackDemo!',
  //   html: `<p>Hello ${username},</p><p>Welcome to DuoStackDemo! We're excited to have you on board.</p>`,
  // });
}

File: backend-dotnet/DuoStackDemo.Api/Services/EmailService.cs
namespace DuoStackDemo.Services;

/// <summary>
/// NoopEmailSender - Placeholder email service for development
/// In production, replace with actual email service (SendGrid, AWS SES, etc.)
/// </summary>
public class EmailService
{
    private readonly JwtService _jwtService;
    private readonly ILogger<EmailService> _logger;

    public EmailService(JwtService jwtService, ILogger<EmailService> logger)
    {
        _jwtService = jwtService;
        _logger = logger;
    }

    /// <summary>
    /// Generate email confirmation token (24h expiry)
    /// </summary>
    public string GenerateConfirmationToken(Guid userId, string email, string username, string role)
    {
        // Generate a token with 24h expiry for email confirmation
        var claims = new[]
        {
            new System.Security.Claims.Claim("userId", userId.ToString()),
            new System.Security.Claims.Claim("purpose", "email-confirmation"),
            new System.Security.Claims.Claim(System.Security.Claims.ClaimTypes.Email, email),
            new System.Security.Claims.Claim(System.Security.Claims.ClaimTypes.Name, username),
            new System.Security.Claims.Claim(System.Security.Claims.ClaimTypes.Role, role)
        };

        return _jwtService.GenerateAccessToken(userId, email, username, role);
    }

    /// <summary>
    /// Verify email confirmation token
    /// </summary>
    public Guid? VerifyConfirmationToken(string token)
    {
        var principal = _jwtService.ValidateToken(token);
        return principal != null ? _jwtService.GetUserIdFromClaims(principal) : null;
    }

    /// <summary>
    /// Send confirmation email (placeholder - logs to console)
    /// </summary>
    public Task SendConfirmationEmailAsync(string email, string token)
    {
        _logger.LogInformation("============================================");
        _logger.LogInformation("📧 EMAIL CONFIRMATION (Development Mode)");
        _logger.LogInformation("============================================");
        _logger.LogInformation("To: {Email}", email);
        _logger.LogInformation("Confirmation Link: http://localhost:3001/confirm-email?token={Token}", token);
        _logger.LogInformation("============================================");
        _logger.LogInformation("⚠️  In production, integrate with SendGrid, AWS SES, or similar service");
        _logger.LogInformation("");

        return Task.CompletedTask;
    }

    /// <summary>
    /// Send password reset email (placeholder - logs to console)
    /// </summary>
    public Task SendPasswordResetEmailAsync(string email, string token)
    {
        _logger.LogInformation("============================================");
        _logger.LogInformation("🔐 PASSWORD RESET (Development Mode)");
        _logger.LogInformation("============================================");
        _logger.LogInformation("To: {Email}", email);
        _logger.LogInformation("Reset Link: http://localhost:3001/reset-password?token={Token}", token);
        _logger.LogInformation("============================================");
        _logger.LogInformation("");

        return Task.CompletedTask;
    }

    /// <summary>
    /// Send welcome email (placeholder - logs to console)
    /// </summary>
    public Task SendWelcomeEmailAsync(string email, string username)
    {
        _logger.LogInformation("============================================");
        _logger.LogInformation("👋 WELCOME EMAIL (Development Mode)");
        _logger.LogInformation("============================================");
        _logger.LogInformation("To: {Email}", email);
        _logger.LogInformation("Username: {Username}", username);
        _logger.LogInformation("Welcome to DuoStackDemo!");
        _logger.LogInformation("============================================");
        _logger.LogInformation("");

        return Task.CompletedTask;
    }
}
